### PR TITLE
Memoize `torch.cuda.device_count`

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84879
* __->__ #84878
* #84877

### CODE REVIEW

File: torch/cuda/__init__.py [hunk 1]
```diff
@@ -15,6 +15,7 @@
 import traceback
 import warnings
 import threading
+from functools import lru_cache as _lru_cache
```
Comment (kit1980) [inline torch/cuda/__init__.py:18]: Why rename?

**Comment by ezyang** [comment]
I'm going to defer to ngimel on this one

File: torch/cuda/__init__.py (hunk 1 above)
Comment (malfet) [reply torch/cuda/__init__.py:18]: I want to be clear, that this is internal implementation detail and not expected to be used outside of this module, where `_` prefix is a pretty good indicator.

Comment (kit1980) [reply torch/cuda/__init__.py:18]: But we don't do `import warnings as _warnings`, for example.

Comment (malfet) [reply torch/cuda/__init__.py:18]: Imo, this would be a good practice. Or use `__all__`

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**